File: 8-agent-to-agent/typescript/src/activities.ts
/**
 * Exercise 8 - Agent to Agent (TypeScript)
 *
 * TODO: This is a placeholder. The Java version of this exercise is fully
 * implemented. See the java/ directory for the complete implementation.
 *
 * The activities needed for the ReAct agent workflow include:
 *   - thoughtActivity: LLM decides next action or final answer
 *   - actionActivity: execute a tool (including A2A calls)
 *   - observationActivity: distill tool results into context
 *   - compactActivity: summarize context when it grows too large
 *   - persistActivity: save messages for history
 */

import dotenv from 'dotenv'
dotenv.config()

export async function thoughtActivity(): Promise<string> {
  throw new Error('Not Implemented — see java/ for reference implementation')
}

export async function actionActivity(): Promise<string> {
  throw new Error('Not Implemented — see java/ for reference implementation')
}

export async function observationActivity(): Promise<string> {
  throw new Error('Not Implemented — see java/ for reference implementation')
}

File: 8-agent-to-agent/java/src/main/java/bitovi/activities/Activities.java
package bitovi.activities;

import java.util.List;

import bitovi.activities.types.ActionInput;
import bitovi.activities.types.CompactResponse;
import bitovi.activities.types.ObservationResponse;
import bitovi.activities.types.PersistMessage;
import bitovi.activities.types.ThoughtResponse;
import io.temporal.activity.ActivityInterface;
import io.temporal.activity.ActivityMethod;
import io.temporal.failure.ApplicationFailure;

/**
 * Activity interface for the ReAct agent workflow.
 *
 * Each method maps to one step in the reasoning cycle:
 *   thoughtActivity     — LLM decides: answer the user or invoke a tool
 *   actionActivity      — execute the chosen tool (including A2A calls)
 *   observationActivity — LLM distills raw tool output into concise context
 *   compactActivity     — summarize context when it grows too large
 *   persistActivity     — save messages for audit/history
 *   getTokenUsage       — estimate token count for compaction threshold
 */
@ActivityInterface
public interface Activities {
	@ActivityMethod
	ThoughtResponse thoughtActivity(List<String> context) throws ApplicationFailure;

	@ActivityMethod
	String actionActivity(String toolName, ActionInput input) throws ApplicationFailure;

	@ActivityMethod
	ObservationResponse observationActivity(String thought, String actionName, String actionInputs, String actionResult) throws ApplicationFailure;

	@ActivityMethod
	CompactResponse compactActivity(List<String> context) throws ApplicationFailure;

	@ActivityMethod
	void persistActivity(List<PersistMessage> messages) throws ApplicationFailure;

	@ActivityMethod
	Integer getTokenUsage(List<String> context) throws ApplicationFailure;
}
